_posts/2022-1-19-quantization-in-practice.md

```diff
@@ -0,0 +1,199 @@
+
+```python
+import torch
+from torch import nn
+
+# toy model
+m = nn.Sequential(
+  nn.Conv1d(2, 64, (8,)),
+  nn.ReLU(),
+  nn.Linear(16,10),
+  nn.LSTM(10, 10))
+
+m.eval()
+
+## EAGER MODE
+from torch.quantization import quantize_dynamic
+model_quantized = quantize_dynamic(
+    model=m, qconfig_spec={nn.LSTM, nn.Linear}, dtype=torch.qint8, inplace=False
+)
+
+## FX MODE
+from torch.quantization import quantize_fx
+qconfig_dict = {"": torch.quantization.default_dynamic_qconfig}  # An empty key denotes the default applied to all modules
+model_prepared = quantize_fx.prepare_fx(m, qconfig_dict)
+model_quantized = quantize_fx.convert_fx(model_prepared)
+```
+
+### Post-Training Static Quantization
+Similar to dynamic quantization, weights are pre-quantized. In dynamic, the activations are calibrated and quantized at inference time. In static, activations are precalibrated by passing sample inputs to the model. 
+
+This method has faster inference than dynamic quantization, but is less robust to accuracy drops from out-of-distribution inputs, i.e. if the model encounters data different from the sample inputs it has calibrated.
+
+```python
+# Static quantization of a model consists of the following steps:
+
+#     Fuse modules
+#     Insert Quant/DeQuant Stubs
+#     Prepare the fused module (insert observers before and after)
+#     Calibrate the prepared module (pass it representative data)
+#     Convert the calibrated module (replace with quantized version)
+
+import torch
+from torch import nn
+
+m = nn.Sequential(
+     nn.Conv1d(2,64,(8,)),
+     nn.ReLU(),
+     nn.Conv1d(64, 128, (1,)),
+     nn.ReLU()
+)
+
+## EAGER MODE
+"""Fuse"""
+torch.quantization.fuse_modules(m, ['0','1'], inplace=True) # fuse first Conv-ReLU pair
+torch.quantization.fuse_modules(m, ['2','3'], inplace=True) # fuse second Conv-ReLU pair
+
+"""Insert stubs"""
+m = nn.Sequential(torch.quantization.QuantStub(), 
+                  *m, 
+                  torch.quantization.DeQuantStub())
+
+"""Prepare"""
+m.qconfig = torch.quantization.get_default_qconfig('fbgemm')
+torch.quantization.prepare(m, inplace=True)
+
+"""Calibrate"""
+with torch.inference_mode():
+  for _ in range(10):
+    x = torch.rand(1,2,28)
+    m(x)
+    
+"""Convert"""
+torch.quantization.convert(m, inplace=True)
+
+"""Check"""
+print(m[1].weight().element_size()) # 1 instead of 4 for FP32
+
+
+## FX GRAPH
+from torch.quantization import quantize_fx
+qconfig_dict = {"": torch.quantization.get_default_qconfig('fbgemm')}
+model_to_quantize.eval()
+# Prepare
+model_prepared = quantize_fx.prepare_fx(model_to_quantize, qconfig_dict)
+# Calibrate 
+with torch.inference_mode():
+  for _ in range(10):
+    x = torch.rand(1,2,28)
+    model_prepared(x)
+# quantize
+model_quantized = quantize_fx.convert_fx(model_prepared)
+```
+
+### Quantization-aware Training (QAT)
```

Comment:
NVIDIA had a nice diagram on QAT since this flow is a bit complex would help to explain in a picture https://developer.nvidia.com/blog/improving-int8-accuracy-using-quantization-aware-training-and-tao-toolkit/

Same comment for the other 2 techniques